src/Std/Sat/AIG/CachedGates.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Std.Sat.AIG.Cached
+import Std.Sat.AIG.CachedLemmas
+
+/-!
+This module contains functions to construct basic logic gates while making use of the sub-circuit
+cache if possible.
+-/
+
+namespace Std
+namespace Sat
+
+namespace AIG
+
+variable {α : Type} [Hashable α] [DecidableEq α]
+
+/--
+Create a not gate in the input AIG. This uses the builtin cache to enable automated subterm sharing.
+-/
+def mkNotCached (aig : AIG α) (gate : Ref aig) : Entrypoint α :=
+  -- ¬x = true && invert x
+  let res := aig.mkConstCached true
+  let aig := res.aig
+  let constRef := res.ref
+  aig.mkGateCached
+    {
+      lhs := {
+        ref := constRef
+        inv := false
+      }
+      rhs := {
+        ref := gate.cast <| by
+          intros
+          apply LawfulOperator.le_size_of_le_aig_size (f := mkConstCached)
+          omega
+        inv := true
+      }
+    }
```

Comment:
```suggestion
    { lhs := 
      { ref := constRef
        inv := false }
      rhs := 
      { ref := gate.cast <| by
          intros
          apply LawfulOperator.le_size_of_le_aig_size (f := mkConstCached)
          omega
        inv := true } }
```
I guess we don't have a definitive style rule on this. Mathlib does it this way...

Reply:
FWIW, I looked at this a few weeks ago and came to the conclusion that Lean uses the original style quite consistently, while mathlib exclusively uses the style you suggested. So I believe that it would be more consistent to format it the way Henrik had it originally.

Reply:
I'm happy to change it either way, just need a clear decision^^.

Reply:
I think there is no clear consensus in the existing Lean codebase at the moment (including mixing different styles within the same style), so I think in the interest of making progress I suggest leaving things as you have them now and not losing too much time on this as long as they adhere to one of these styles. However, neither style should have an opening brace on its own line (so [this](https://github.com/leanprover/lean4/pull/4953/files#diff-86e39419197a4029f16533cec6041c2f9c1d08882e2e4007609f78a06b27620cR72-R73) is not okay), and within a declaration you should stick to one style (so [this](https://github.com/leanprover/lean4/pull/4953/files#diff-05f9943490c5a32907d1aec86f0c49bc9002579f8c383a574f8afaf6463a0fb8R29-R37) is not okay).